File: src/java/magma/app/compile/rule/or/OrRule.java
package magma.app.compile.rule.or;

import magma.api.result.Result;
import magma.app.compile.context.Context;
import magma.app.compile.context.NodeContext;
import magma.app.compile.context.StringContext;
import magma.app.compile.error.FormattedError;
import magma.app.compile.node.DisplayNode;
import magma.app.compile.result.ResultCreator;
import magma.app.compile.rule.Rule;

import java.util.List;
import java.util.function.Function;

public final class OrRule<Node extends DisplayNode> implements Rule<Node, Result<Node, FormattedError>, Result<String, FormattedError>> {
    private final List<Rule<Node, Result<Node, FormattedError>, Result<String, FormattedError>>> rules;
    private final ResultFactory<Node, Result<Node, FormattedError>, Result<String, FormattedError>> factory;

    public OrRule(List<Rule<Node, Result<Node, FormattedError>, Result<String, FormattedError>>> rules, ResultFactory<Node, Result<Node, FormattedError>, Result<String, FormattedError>> factory) {
        this.rules = rules;
        this.factory = factory;
    }

    @Override
    public Result<Node, FormattedError> lex(String input) {
        return this.or(this.factory.createNodeCreator(), rule1 -> rule1.lex(input), new StringContext(input));
    }

    private <Value> Result<Value, FormattedError> or(ResultCreator<Value, Result<Value, FormattedError>> factory, Function<Rule<Node, Result<Node, FormattedError>, Result<String, FormattedError>>, Result<Value, FormattedError>> mapper, Context context) {
        return this.rules.stream()
                .<OrState<Value, FormattedError, Result<Value, FormattedError>>>reduce(new MutableOrState<>(factory),
                        (state, rule) -> mapper.apply(rule)
                                .match(state::withValue, state::withError),
                        (_, next) -> next)
                .toResult(context);
    }

    @Override
    public Result<String, FormattedError> generate(Node node) {
        return this.or(this.factory.createStringCreator(), rule1 -> rule1.generate(node), new NodeContext(node));
    }
}

File: src/clang/magma/compile/rule/tree/OrRule.c
#include "OrRule.h"
magma.result.Result<magma.compile.rule.tree.T, magma.compile.CompileError> apply(magma.result.Result<magma.compile.rule.tree.T, magma.compile.CompileError>(*applicator)(magma.compile.rule.Rule), magma.compile.context.Context(*context)()){return rules.stream().foldWithInitial((), __lambda0__.apply(rule).match(orState.withValue, orState.withError)).toResult().match(Ok.new, __lambda1__);
}
magma.result.Result<magma.compile.Node, magma.compile.CompileError> parse(String input){return apply(__lambda2__.parse(input), __lambda3__);
}
magma.result.Result<String, magma.compile.CompileError> generate(magma.compile.Node input){return apply(__lambda4__.generate(input), __lambda5__);
}
auto __lambda0__();
auto __lambda1__();
auto __lambda2__();
auto __lambda3__();
auto __lambda4__();
auto __lambda5__();

